File: src/instructions/tools/git-report/git-report.py
#!/usr/bin/env python3
"""git-report — snapshot current git state and output markdown or JSON.

Captures: current branch, default remote branch, recent commits, working-tree
status, staged diff (summary + full), and commits ahead of origin/HEAD.

Usage:
  python3 git-report.py
  python3 git-report.py --format markdown
  python3 git-report.py --log-count 20
"""
import argparse
import json
import subprocess
import sys


def run(cmd: list[str], cwd: str | None = None) -> tuple[str, int]:
    """Run a shell command, return (stdout, returncode). Never raises."""
    try:
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=30,
            cwd=cwd,
        )
        return result.stdout.strip(), result.returncode
    except subprocess.TimeoutExpired:
        return "(timeout)", 1
    except FileNotFoundError:
        return "(git not found)", 1


def collect(log_count: int) -> dict:
    """Collect all git state into a structured dict."""
    # Ensure origin/HEAD is resolved (best-effort, ignore failure)
    run(["git", "remote", "set-head", "origin", "--auto"])

    current_branch, _ = run(["git", "branch", "--show-current"])
    default_branch, rc = run(["git", "symbolic-ref", "--short", "refs/remotes/origin/HEAD"])
    if rc != 0:
        default_branch = "(unknown — run: git remote set-head origin --auto)"

    recent_commits, _ = run(["git", "log", "--oneline", f"-{log_count}"])
    status, _ = run(["git", "status"])
    staged_stat, _ = run(["git", "diff", "--staged", "--stat"])
    staged_diff, _ = run(["git", "diff", "--staged"])

    ahead_commits = ""
    if "unknown" not in default_branch:
        ahead_commits, _ = run(
            ["git", "log", "--oneline", "HEAD", f"^{default_branch}", "--no-merges"]
        )

    fsck_out, _ = run(["git", "fsck", "--full"])
    fsck_missing = "\n".join(
        line for line in fsck_out.splitlines() if "missing" in line
    )

    return {
        "current_branch": current_branch,
        "default_branch": default_branch,
        "recent_commits": recent_commits,
        "status": status,
        "staged_stat": staged_stat,
        "staged_diff": staged_diff,
        "ahead_commits": ahead_commits,
        "fsck_missing": fsck_missing,
    }


def render_markdown(data: dict) -> str:
    """Render collected git state as a markdown report."""
    lines = []

    lines.append(f"## Git Report")
    lines.append("")
    lines.append(f"**Current branch:** `{data['current_branch']}`  ")
    lines.append(f"**Default remote branch:** `{data['default_branch']}`")
    lines.append("")

    lines.append("### Recent Commits")
    lines.append("")
    if data["recent_commits"]:
        for line in data["recent_commits"].splitlines():
            lines.append(f"- {line}")
    else:
        lines.append("_(no commits)_")
    lines.append("")

    lines.append("### Status")
    lines.append("")
    lines.append("```")
    lines.append(data["status"] or "(clean)")
    lines.append("```")
    lines.append("")

    lines.append("### Staged Changes (summary)")
    lines.append("")
    if data["staged_stat"]:
        lines.append("```")
        lines.append(data["staged_stat"])
        lines.append("```")
    else:
        lines.append("_(nothing staged)_")
    lines.append("")

    lines.append("### Staged Changes (full diff)")
    lines.append("")
    if data["staged_diff"]:
        lines.append("```diff")
        lines.append(data["staged_diff"])
        lines.append("```")
    else:
        lines.append("_(nothing staged)_")
    lines.append("")

    lines.append("### Commits Ahead of Default Branch")
    lines.append("")
    if data["ahead_commits"]:
        for line in data["ahead_commits"].splitlines():
            lines.append(f"- {line}")
    else:
        lines.append("_(none — branch is up to date or default branch unknown)_")
    lines.append("")

    lines.append("### Index Integrity (fsck missing objects)")
    lines.append("")
    if data["fsck_missing"]:
        lines.append("```")
        lines.append(data["fsck_missing"])
        lines.append("```")
    else:
        lines.append("_(no missing objects — index clean)_")

    return "\n".join(lines)


def main() -> None:
    parser = argparse.ArgumentParser(description="git-report tool")
    parser.add_argument(
        "--log-count",
        type=int,
        default=10,
        help="Number of recent commits to show (default: 10)",
    )
    parser.add_argument(
        "--format",
        choices=["markdown", "json"],
        default="markdown",
        help="Output format (default: markdown)",
    )
    args = parser.parse_args()

    data = collect(args.log_count)

    if args.format == "json":
        print(json.dumps({"status": "ok", **data}, indent=2))
    else:
        print(render_markdown(data))


if __name__ == "__main__":
    main()

File: src/instructions/tools/git-report/git-report.ts
import { tool } from "@opencode-ai/plugin"
import path from "path"
import os from "os"
import fs from "fs"

/**
 * git-report — snapshot current git state.
 *
 * Args (no Zod schema — args:{} to avoid cross-instance crash, see opencode gotcha):
 *   log_count  number  optional  Number of recent commits (default: 10)
 *   format     string  optional  "json" (default) or "markdown"
 */

export default tool({
  description:
    "Return a git state snapshot: current branch, default remote branch, recent commits, working-tree status, staged diff (summary + full), commits ahead of origin/HEAD, and index integrity check (missing objects from git fsck). Use this to understand what has changed before committing, reviewing, or planning next steps.\n\nParameters:\n- log_count (number, optional): Number of recent commits to show (default: 10)\n- format (string, optional): 'json' (default) or 'markdown'",
  args: {},
  async execute(rawArgs) {
    const args = rawArgs as any
    const script = path.join(import.meta.dir, "git-report.py")
    const fmt = (args.format as string | undefined) ?? "json"

    const cmd = [
      "python3",
      script,
      "--format",
      fmt,
      "--log-count",
      String((args.log_count as number | undefined) ?? 10),
    ]

    const proc = Bun.spawn(cmd, { stdio: ["ignore", "pipe", "pipe"] })
    const stdout = await new Response(proc.stdout).text()
    const stderr = await new Response(proc.stderr).text()
    const exitCode = await proc.exited

    if (exitCode !== 0) {
      return `git-report failed (exit ${exitCode})\n${stderr.trim() || stdout.trim() || "(no output)"}`
    }

    if (fmt === "json") {
      return stdout.trim()
    }

    // Format markdown tables before returning
    const tmpFile = path.join(os.tmpdir(), `git-report-${Date.now()}.md`)
    try {
      await Bun.write(tmpFile, stdout.trim())
      const fmtScript = path.join(import.meta.dir, "..", "format-md", "format-md.py")
      const fmtProc = Bun.spawn(["python3", fmtScript, tmpFile], { stdio: ["ignore", "pipe", "pipe"] })
      await fmtProc.exited
      return (await Bun.file(tmpFile).text()).trim()
    } finally {
      try { fs.unlinkSync(tmpFile) } catch { /* ignore */ }
    }
  },
})
